Kotlin conversion of AboutActivity with tests also in Kotlin

juuce79 opened the pull request:
- Converted `AboutActivity` to kotlin from java.
- Have tried to keep the code as close to the original as possible while using kotlin code instead of java code.
- Added tests for AboutActivity.
- Manually tested with APIs 24, 30 and 35 in Virtual Device in Android Studio.
- This is part of my university project, to contribute to an open source project, and part of developer's plan to convert app to firstly Kotlin and eventually Jetpack Compose

theimpulson commented:
Related PRs that should be closed if this supersedes them:

- https://github.com/CatimaLoyalty/Android/pull/1122
- https://github.com/CatimaLoyalty/Android/pull/1176

theimpulson commented on app/src/main/java/protect/card_locker/AboutActivity.kt:
```diff
@@ -0,0 +1,146 @@
+            OpenWebLinkHandler().openBrowser(this, tag)
+        }
+    }
+}
```
Missing new line EOF

theimpulson commented on app/src/main/java/protect/card_locker/AboutActivity.kt:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        if (item.itemId == android.R.id.home) {
+            finish()
+        }
+        return super.onOptionsItemSelected(item)
```
You can consider using a `when` block here for a cleaner and easily extensible code.

```suggestion
        return when (item.itemId) {
            android.R.id.home -> finish()
            else -> super.onOptionsItemSelected(item)
        }
```

juuce79 replied:
updated with @theimpulson suggestion

theimpulson commented on app/src/main/java/protect/card_locker/AboutActivity.kt:
```diff
@@ -0,0 +1,146 @@
+            rate.tag = "https://play.google.com/store/apps/details?id=me.hackerchick.catima"
+            donate.tag = "https://catima.app/donate"
+
+            // Hide Google Play rate button if not on Google Play
+            rate.visibility = if (BuildConfig.showRateOnGooglePlay) View.VISIBLE else View.GONE
+            // Hide donate button on Google Play (Google Play doesn't allow donation links)
+            donate.visibility = if (BuildConfig.showDonate) View.VISIBLE else View.GONE
```
You can use the [isVisible](https://developer.android.com/reference/kotlin/androidx/core/view/package-summary#(android.view.View).isVisible()) extension method from the `core-ktx` library here to make the logic more clean and simple.

juuce79 replied:
updated with @theimpulson suggestion

theimpulson commented on app/src/main/java/protect/card_locker/AboutActivity.kt:
```diff
@@ -0,0 +1,146 @@
+        private const val TAG = "Catima"
+    }
+
+    private var _binding: AboutActivityBinding? = null
+    private val binding get() = _binding!!
```
This pattern of view binding is used for fragments and not activities. Please see [this](https://developer.android.com/topic/libraries/view-binding#activities) guide on patterns for activities.

juuce79 replied:
updated with @theimpulson suggestion

theimpulson commented on app/src/main/java/protect/card_locker/AboutActivity.kt:
```diff
@@ -0,0 +1,146 @@
+        return super.onOptionsItemSelected(item)
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        content.destroy()
+        clearClickListeners()
+        _binding = null
+    }
```
These methods are not required for an Activity and should be dropped. They are only relevant for Fragment.

juuce79 replied:
updated with @theimpulson suggestion

juuce79 commented:
I have updated the code with your suggestions and I was wondering, can I update this pull request with a new commit of the updated code (I haven't pushed or committed it to my fork yet so I don't mess anything up), or do I need to make a new pull request?
Regarding this code
```
    override fun onDestroy() {
        super.onDestroy()
        content.destroy()
        clearClickListeners()
        _binding = null
    }
```
I have just removed the `_binding = null`.

As this is being done as part of my university program and I am documenting the process, here is a PDF of the conversion process with the new updated code included:
[catima_kotlin_conversion (2).pdf](https://github.com/user-attachments/files/19004443/catima_kotlin_conversion.2.pdf)

Thanks for the help by the way, I and my partner (we are 2 working on this project) are still at university and have only had one Kotlin course so we are still new to Kotlin, and Java to be honest, although we have had quite a few Java courses so far. We are currently in our third year of our bachelors degree.

juuce79 commented:
Updated with Aayush Gupta's suggestions:
- added new line at EOF.
- changed if-else statement to `when` block for cleaner and more easily extensible code (lines 57-63).
- now using `isVisible` extension method from the core-ktx library to make the logic more clean and simple (lines 48 & 50).
- now using correct view binding pattern for activities (line 22).
- removed `_binding = null` from `onDestroy()` function (line 71).

theimpulson commented on app/src/test/java/protect/card_locker/AboutActivityTest.kt:
```diff
@@ -0,0 +1,174 @@
+        activityController.create().start().resume()
+
+        // Test Google Play rate button visibility based on BuildConfig
+        val rateButton = activity.findViewById<View>(R.id.rate)
+        assertEquals(rateButton.visibility,
+            if (BuildConfig.showRateOnGooglePlay) View.VISIBLE else View.GONE)
+
+        // Test donate button visibility based on BuildConfig
+        val donateButton = activity.findViewById<View>(R.id.donate)
+        assertEquals(donateButton.visibility,
+            if (BuildConfig.showDonate) View.VISIBLE else View.GONE)
```
These could be switched to `isVisible` extension method as well like above for better readability.

juuce79 replied:
updated with @theimpulson's suggestion. Now:
```kotlin
 51    @Test
 52    fun testDisplayOptionsBasedOnConfig() {
 53        activityController.create().start().resume()
 54
 55        // Test Google Play rate button visibility based on BuildConfig
 56        val rateButton = activity.findViewById<View>(R.id.rate)
 57        assertEquals(BuildConfig.showRateOnGooglePlay, rateButton.isVisible)
 58
 59        // Test donate button visibility based on BuildConfig
 60        val donateButton = activity.findViewById<View>(R.id.donate)
 61        assertEquals(BuildConfig.showDonate, donateButton.isVisible)
 62    }
```

theimpulson commented on app/src/test/java/protect/card_locker/AboutActivityTest.kt:
```diff
@@ -0,0 +1,174 @@
+    fun testButtonVisibilityBasedOnBuildConfig() {
+        activityController.create().start().resume()
+
+        // Get the current values from BuildConfig
+        val showRateOnGooglePlay = BuildConfig.showRateOnGooglePlay
+        val showDonate = BuildConfig.showDonate
+
+        // Test that the visibility matches the BuildConfig values
+        assertEquals(if (showRateOnGooglePlay) View.VISIBLE else View.GONE,
+            activity.findViewById<View>(R.id.rate).visibility)
+        assertEquals(if (showDonate) View.VISIBLE else View.GONE,
+            activity.findViewById<View>(R.id.donate).visibility)
```
Similar suggestion to use `isVisible` like above

juuce79 replied:
updated with @theimpulson's suggestion. Now:
```kotlin
145    @Test
146    fun testButtonVisibilityBasedOnBuildConfig() {
147        activityController.create().start().resume()
148
149        // Get the current values from BuildConfig
150        val showRateOnGooglePlay = BuildConfig.showRateOnGooglePlay
151        val showDonate = BuildConfig.showDonate
152
153        // Test that the visibility matches the BuildConfig values
154        assertEquals(showRateOnGooglePlay, activity.findViewById<View>(R.id.rate).isVisible)
155        assertEquals(showDonate, activity.findViewById<View>(R.id.donate).isVisible)
156    }
```

juuce79 commented:
Updated test functions testDisplayOptionsBasedOnConfig and `testButtonVisibilityBasedOnBuildConfig` to use `isVisible` instead of checking visibility property directly with `View.VISIBLE/GONE` comparisons. This makes tests more consistent with implementation code in `AboutActivity.kt` and improves test readability.

theimpulson reviewed (approved):
Looks good to me. Thank you.

TheLastProject commented:
Tested it and it works great. Thanks for the rewrite @juuce79 (and also thank you for the PDF file, it made understanding your changes a lot easier!). And thank you for the review, @theimpulson! :)
